envoy/network/address.h

```diff
@@ -230,6 +231,32 @@ class Instance {
    * @return SocketInterface to be used with the address.
    */
   virtual const Network::SocketInterface& socketInterface() const PURE;
+
+  /**
+   * For support the absl::Hash.
+   */
+  virtual void abslHashValue(absl::HashState state) const PURE;
+};
+
+/**
+ * Wrapper for InstanceConstSharedPtr, enable to use InstanceConstSharedPtr
+ * hash key for absl::Hash.
+ */
+struct AddressKey {
+  AddressKey(InstanceConstSharedPtr address) : address_(address) {}
```

Comment:
why do we need the constructor and member variable? it it just for tests because it appears to not be used outside of them.  I'd prefer to avoid it, but if it's unavoidable at least comment it's test only?

Reply:
Not sure I understand your question. The AddressKey is a wrap of `InstanceConstSharedPtr`, I want to store the shared ptr in the key, not the object value that is too expensive, and I can't override the `operator==` for shared ptr. That is why I added the AddressKey. The member variable `address_` is the actual thing as the key. The constructor can implicitly convert `InstanceConstSharedPtr` to the AddressKey, which is more convenient. Let me know if I didn't answer your question.

Reply:
I don't think `AddressKey` is necessary--absl containers support heterogenous lookup for this sort of use case. See https://abseil.io/tips/144.

IIUC (please confirm), the containers that store `InstanceConstSharedPtr` are fundamentally sets, but we need to be able to compute hashes, compare equality, and do lookups based on a subset of the pointee's state. Using maps instead of sets is one way of handling that, but I think it would be more efficient and cleaner to use sets with heterogenous lookup enabled.

Reply:
Emm...do you mean we can hash on `InstanceConstSharedPtr` directly? Is that means we hash on the memory address? If yes, is that safe?

Reply:
What @toddmgreer is suggesting is that you store the `InstanceConstSharedPtr` in the hash map and add custom hash and comparator classes in the `absl::flat_hash_map` template parameters. You can move the == operator and the AbslHashValue to these custom classes.

Moreover you can add methods to the comparator class to compare (as well as the require tag) `Network::Address::InstanceConstSharedPtr` with `Network::Address::Instance` and in this way you would not need to change the `getBalancedHandlerByAddress` method parameter type. You'll be bale to make lookups in the hash map by using the `Network::Address::Instance` type. This is what's called heterogeneous lookup in the ling the Todd have provided.

Reply:
Thanks for explaining those to me again. I did a test, it works, I will update the PR. The sample I wrote is as below:
```
#include <iostream>
#include <string>
#include <memory>
#include <assert.h>
#include "absl/container/flat_hash_map.h"

// Base class
class MyType {
public:
	virtual ~MyType() {}
	virtual bool operator==(const MyType& rhs) const = 0;
	virtual int getType() const = 0;
};

// One sub type
class MySubType1 : public MyType {
public:
	MySubType1(int i) : i_(i) {}
	~MySubType1() override {}
	bool operator==(const MyType& rhs) const override {
		auto sub_rhs = dynamic_cast<const MySubType1&>(rhs);
		return i_ == sub_rhs.i_;
	}

	int getType() const override { return 1; }

	template <typename H>
	friend H AbslHashValue(H h, const MySubType1& m);

	int i_;
};

bool operator==(const MySubType1& lhs, const MySubType1& rhs) {
  return lhs.i_ == rhs.i_;
}

template <typename H>
H AbslHashValue(H h, const MySubType1& m) {
  return H::combine(std::move(h), m.i_);
}

// Another sub type.
class MySubType2 : public MyType {
public:
	MySubType2(std::string s) : s_(s) {}
	~MySubType2() override {}
	bool operator==(const MyType& rhs) const override {
		auto sub_rhs = dynamic_cast<const MySubType2&>(rhs);
		return s_ == sub_rhs.s_;
	}

	int getType() const override { return 2; }

	template <typename H>
	friend H AbslHashValue(H h, const MySubType2& m);

	std::string s_;
};

bool operator==(const MySubType2& lhs, const MySubType2& rhs) {
  return lhs.s_ == rhs.s_;
}

template <typename H>
H AbslHashValue(H h, const MySubType2& m) {
  return H::combine(std::move(h), m.s_);
}

using MyTypeSharedPtr = std::shared_ptr<MyType>;

// The custom hash MyTypeSharedPtr 
struct MyTypeHash {
	using is_transparent = void;
	std::size_t operator()(const MyTypeSharedPtr& s) const noexcept
    {
		if (s->getType() == 1) {
			return absl::Hash<MySubType1>{}(dynamic_cast<const MySubType1&>(*s));
		}

        return absl::Hash<MySubType2>{}(dynamic_cast<const MySubType2&>(*s));
    }
	std::size_t operator()(const MyType& s) const noexcept
    {
        if (s.getType() == 1) {
			return absl::Hash<MySubType1>{}(dynamic_cast<MySubType1 const&>(s));
		}

        return absl::Hash<MySubType2>{}(dynamic_cast<MySubType2 const&>(s));
    }
};

// The custom comparator for MyTypeSharedPtr 
struct MyTypeEqual {
 using is_transparent = void;

 bool operator()(const MyTypeSharedPtr& lhs, const MyTypeSharedPtr& rhs) const
 {
    return *lhs == *rhs;
 }

 bool operator()(const MyTypeSharedPtr& lhs, const MyType& rhs) const {
	return *lhs == rhs;
 }

 bool operator()(const MyType& lhs, const MyTypeSharedPtr& rhs) const {
	return lhs == *rhs;
 }
 bool operator()(const MyType& lhs, const MyType& rhs) const {
	return lhs == rhs;
 }
};

int main(int argc, const char** argv) {
	absl::flat_hash_map<MyTypeSharedPtr, int, MyTypeHash, MyTypeEqual> map;

	std::shared_ptr<MyType> t1 = std::make_shared<MySubType1>(1);
	map.insert({t1, 1});

	MyType& tr1 = *t1;
	auto iter = map.find(tr1);
	assert(iter != map.end());
	assert(iter->second == 1);

	std::shared_ptr<MyType> t2 = std::make_shared<MySubType1>(1);
	map.insert_or_assign(t2, 2);
	auto iter2 = map.find(t2);
	assert(iter2->first.get() != t2.get());
	assert(iter2->second == 2);

	std::shared_ptr<MyType> t3 = std::make_shared<MySubType1>(2);
	map.insert_or_assign(t3, 3);
	auto iter3 = map.find(t3);
	assert(iter3->first.get() == t3.get());
	assert(iter3->second == 3);
}
```

Reply:
@toddmgreer Another suggestion from you is using absl::flat_hash_set instead of absl::flat_hash_map, since the value is `PerAddressActiveListenerDetails`, it is include address also, then I can use the heterogenous lookup to query by the address. Did I understand correctly?

One issue is we have other places use the address to hash also, like https://github.com/envoyproxy/envoy/blob/4fb3eeb1a12b9005133639e96425fef16816c2a4/source/extensions/listener_managers/listener_manager/listener_impl.h#L425
https://github.com/envoyproxy/envoy/blob/4fb3eeb1a12b9005133639e96425fef16816c2a4/source/extensions/listener_managers/listener_manager/listener_impl.h#L510

Seems like I should create custom function and comparator for each of them. Is that something we want?